phpBB/phpbb/db/driver/mysqli.php

```diff
@@ -331,7 +331,10 @@ function sql_freeresult($query_id = false)
 	*/
 	function sql_escape($msg)
 	{
-		return @mysqli_real_escape_string($this->db_connect_id, $msg);
+		return @mysqli_real_escape_string(
+			$this->db_connect_id,
+			utf8_encode_ucr($msg)
```

Comment:
This is assuming the intention for all usage and all extensions that already use this function. What exactly happens when this function has already been called?

Reply:
Because `utf8_encode_ucr` only modifies non-BMP unicode characters, and HTML entities are ASCII only (i.e. will always lie well within the BMP), that means it's idempotent for multiple calls. In other words, `utf8_encode_ucr(utf8_encode_ucr(utf8_encode_ucr($str)))` will always yield the same result as `utf8_encode_ucr($str)`.

See this playground on repl.it (functions copied directly from `includes/utf/utf_tools.php`):

https://replit.com/@lionel_rowe/utf8encodeucr

Reply:
Not convinced this is the right approach. It is arbitrarily deciding how everyone should store their data and will likely produce unexpected results. Probably better to do specific implementations where needed.

Reply:
> Probably better to do specific implementations where needed.

As it has been done so far, let's say there are no bugs yet uncovered or there are some areas that have not fallbacks. Everything has been covered until proven otherwise.

Reply:
> arbitrarily deciding how everyone should store their data

I wholeheartedly agree in principle, but unfortunately that ship has sailed long ago, back when `utf8_encode_ucr` was introduced into the codebase in the first place, and probably even before that — the vast majority of string fields in phpBB are already (unnecessarily) stored as HTML, which should really be handled at the templating layer. For example, if I create a user with the username "&lt;jimmy&gt;", this is stored in the database as `&lt;jimmy&gt;`. All this PR does is ensure that (for example) username "jimmy💩" would also be stored as `jimmy&#128169;` to avoid MYSQLi throwing errors all over the place, which I've already encountered multiple times (I use a lot of emojis 😂).

Worst case with this PR is that a user occasionally sees strings looking like "&amp;#128169;", as opposed to the current worst case, which is that they get full-page error messages and potential data loss (e.g. the action they were performing is lost).

If this was merged along with https://github.com/phpbb/phpbb/pull/6377, even that relatively minor tradeoff could also be eliminated.

Reply:
> that ship has sailed long ago, back when utf8_encode_ucr was introduced into the codebase in the first place

That doesn't make any sense. Are you saying that just because it's there it should be applied everywhere? Don't you think that would have been done back then?

Reply:
> That doesn't make any sense. Are you saying that just because it's there it should be applied everywhere?

I'm saying that if the de-facto normalized form for everything is HTML + encoded non-BMP codepoints, then it makes sense to normalize everything to that de-facto standard, even if we wish the standard had been something else (such as pure plain-text, with a database supporting full UTF-8 as a system requirement).

> Don't you think that would have been done back then?

I think there are many less-than-ideal things about the phpBB codebase, as surely everyone will agree. That's natural and to be expected on an open-source project of this size and age and with this many contributors. All I'm trying to do is find some small ways to improve consistency and eliminate certain classes of recurring bugs.

Reply:
> I think there are many less-than-ideal things about the phpBB codebase, as surely everyone will agree.

Ok, you're putting new logic into a function that is core to the product (dbal) and have so far justified it with "the function exists so it should be used in a spot that affects any and all code that uses it (that ship has sailed)", "if it breaks someone's code then so be it (user will sometimes see HTML encoded representations of the text)", and "the function wasn't used here because the code base is less than ideal".

While these aren't really good reasons/arguments I realize this PR is for master anyways so it's probably ok as a potentially breaking change for the next major version given the expectation that some extensions may need some changes anyways.

Reply:
@DavidIQ Do you think it is a good idea to have Emoji in usernames or other parts of the code? I don't think so as we have already discussed this and put fallbacks in place.

Reply:
> if I create a user with the username "<jimmy>", this is stored in the database as `&lt;jimmy&gt;`. All this PR does is ensure that (for example) username "jimmy💩" would also be stored as `jimmy&#128169;` to avoid MYSQLi throwing errors all over the place

It does not make sense, please try first. There are multiple configurations for usernames at registration time. I would suggest that you learn more about phpBB at the UI level. Names like `<jimmy>` are possible.

Reply:
@3D-I the username thing was just an example. Also, if an admin wants to allow users to have emojis in their username, why shouldn't that be allowed?